### PR TITLE
Add DialUDP and ResolveUDP

### PR DESCRIPTION
**Pull request by atavism**

### CODE REVIEW

File: protected.go [hunk 1]
```diff
@@ -54,28 +69,71 @@ func New(protect Protect, dnsServerIP string) *Protector {
 // Resolve resolves the given address using a DNS lookup on a UDP socket
 // protected by the given Protect function.
 func (p *Protector) Resolve(network string, addr string) (*net.TCPAddr, error) {
-	op := ops.Begin("protected-resolve").Set("addr", addr)
```
Comment (myleshorton) [inline protected.go]: Do we want to ditch `ops` here @atavism?

Comment (atavism) [reply protected.go]: @myleshorton I just moved the `ops` inside `resolve`

File: protected.go [hunk 2]
```diff
@@ -54,28 +69,71 @@ func New(protect Protect, dnsServerIP string) *Protector {
...
-	defer op.End()
-	conn, err := p.resolve(op, network, addr)
-	return conn, op.FailIf(err)
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+	case "udp", "udp4", "udp6":
```
Comment (oxtoacart) [inline protected.go]: Why are we allowing udp here?

File: protected.go (hunk 1 above)
Comment (oxtoacart) [inline protected.go]: By moving this down into `resolve`, we lose the ability to mark this op as failed when we get the unsupported network error.

File: protected.go [hunk 3]
```diff
@@ -54,28 +69,71 @@ func New(protect Protect, dnsServerIP string) *Protector {
...
+func (p *Protector) ResolveUDP(network, addr string) (*net.UDPAddr, error) {
+	switch network {
+	case "udp", "udp4", "udp6":
+		break
+	default:
+		err := errors.New("ResolveUDP: Unsupported network: %s", network)
+		log.Error(err)
+		return nil, err
+	}
+	conn, err := p.resolve(network, addr)
```
Comment (oxtoacart) [inline protected.go]: `conn` is kind of a misnomer

File: protected.go [hunk 4]
```diff
@@ -117,14 +179,19 @@ func (p *Protector) resolve(op ops.Op, network string, addr string) (*net.TCPAdd
...
+	return getProtectedAddr(network, ipAddr, port)
 }
 
 // Dial creates a new protected connection.
 // - syscall API calls are used to create and bind to the
 //   specified system device (this is primarily
 //   used for Android VpnService routing functionality)
-func (p *Protector) Dial(network, addr string, timeout time.Duration) (net.Conn, error) {
+func (p *Protector) Dial(network, addr string) (net.Conn, error) {
+	ctx, _ := context.WithTimeout(context.Background(), defaultDialTimeout)
```
Comment (oxtoacart) [inline protected.go]: We need to call the returned cancel() function in a defer.

File: protected.go (hunk 2 above)
Comment (atavism) [reply protected.go]: We shouldn't -- I think that was a leftover from when I was debugging. Just removed that line.

File: protected.go (hunk 3 above)
Comment (atavism) [reply protected.go]: Renamed it

File: protected.go (hunk 2 above)
Comment (oxtoacart) [reply protected.go]: Oh, got it, could be either

File: protected.go (hunk 4 above)
Comment (atavism) [reply protected.go]: Yeah, interesting.. that isn't a change I made (guess it's been that way awhile). Just updated it.

**Pull request merged**